### PR TITLE
feat: inline methods consisting of a single return statement or local variable declaration expression

### PR DESCRIPTION
**Pull request by hartmair**
# feat: inline methods consisting of a single return statement or local variable declaration expression

## Description

- feat: inline methods consisting of a single return statement
- feat: inline methods consisting of a single local variable declaration

Fixes # (issue)

## Checklist

- [x] The existing code style is followed
- [x] The commit message follows our guidelines
- [x] Performed a self-review of my code
- [ ] Hard-to-understand areas of my code are commented
- [x] The documentation is updated (as applicable)
- [x] Unit tests are added/updated
- [ ] Integration tests are added/updated (as applicable, especially if feature/bug depends on roslyn or framework version in use)

### CODE REVIEW

**Comment by trejjam** [comment]
You can do this with current features:

Source:
```cs
[Mapper]
public static partial class Mapper
{
    private static partial System.Linq.IQueryable<B> Map(this System.Linq.IQueryable<A> source);

    [MapProperty(nameof(C.Value), nameof(D.Value), Use = nameof(GetComputedValue))]
    private static partial D MapToD(C v);

    [UserMapping(Default = false)]
    private static string GetComputedValue(string value) => value + "-mapped";
}

class A { public string StringValue { get; set; } public C NestedValue { get; set; } }
class B { public string StringValue { get; set; } public D NestedValue { get; set; } }
class C { public string Value { get; set; } }
class D { public string Value { get; set; } }
```

Generated
```cs
// <auto-generated />
#nullable enable


    public static partial class Mapper
    {
        [global::System.CodeDom.Compiler.GeneratedCode("Riok.Mapperly", "0.0.1.0")]
        private static partial global::System.Linq.IQueryable<B> Map(this global::System.Linq.IQueryable<A> source)
        {
#nullable disable
            return System.Linq.Queryable.Select(source, x => new B()
            {
                StringValue = x.StringValue,
                NestedValue = new D()
                {
                    Value = x.NestedValue.Value + "-mapped",
                },
            });
#nullable enable
        }

        [global::System.CodeDom.Compiler.GeneratedCode("Riok.Mapperly", "0.0.1.0")]
        private static partial D MapToD(C v)
        {
            var target = new D();
            target.Value = GetComputedValue(v.Value);
            return target;
        }
    }
```

**Comment by hartmair** [comment]
@trejjam I know that you can do this if you use an expression body. That's the point of this pull request. That you don't have to think about the code style. It just works regardlessly how you write your user mapping.

File: src/Riok.Mapperly/Descriptors/MappingBuilders/InlineExpressionMappingBuilder.cs [hunk 1]
```diff
@@ -25,21 +25,51 @@ public static class InlineExpressionMappingBuilder
         }
 
         var methodSyntax = methodSyntaxRef.GetSyntax();
-        if (methodSyntax is not MethodDeclarationSyntax { ExpressionBody: { } body, ParameterList.Parameters: [var sourceParameter] })
+
+        var bodyExpression = methodSyntax switch
+        {
+            // => expression
+            MethodDeclarationSyntax { ExpressionBody: { } body, ParameterList.Parameters: [var sourceParameter1] } => body.Expression,
```
Comment (latonz) [inline src/Riok.Mapperly/Descriptors/MappingBuilders/InlineExpressionMappingBuilder.cs:32]: `ParameterList.Parameters: [var sourceParameter1]` could be simplified to `ParameterList.Parameters.Count: 1`

File: src/Riok.Mapperly/Descriptors/MappingBuilders/InlineExpressionMappingBuilder.cs [hunk 2]
```diff
@@ -25,21 +25,51 @@ public static class InlineExpressionMappingBuilder
...
+            // { return expression; }
+            MethodDeclarationSyntax { Body.Statements: [ReturnStatementSyntax singleStatement] } => singleStatement.Expression,
+
+            // { var dest = expression; return dest; }
+            MethodDeclarationSyntax
+            {
+                Body.Statements: [
+                    LocalDeclarationStatementSyntax
+                    {
+                        Declaration.Variables: [VariableDeclaratorSyntax { Initializer: { } variableInitializer } variableDeclarator]
```
Comment (latonz) [inline src/Riok.Mapperly/Descriptors/MappingBuilders/InlineExpressionMappingBuilder.cs:43]: `VariableDeclaratorSyntax` is redundant

File: src/Riok.Mapperly/Descriptors/MappingBuilders/InlineExpressionMappingBuilder.cs [hunk 3]
```diff
@@ -25,21 +25,51 @@ public static class InlineExpressionMappingBuilder
...
         {
             ctx.ReportDiagnostic(DiagnosticDescriptors.QueryableProjectionMappingCannotInline, mapping.Method);
             return null;
         }
 
+        if (methodSyntax is not MethodDeclarationSyntax { ParameterList.Parameters: [var sourceParameter] })
+        {
+            ctx.ReportDiagnostic(DiagnosticDescriptors.QueryableProjectionMappingCannotInline, mapping.Method);
+            return null;
+        }
```
Comment (latonz) [inline src/Riok.Mapperly/Descriptors/MappingBuilders/InlineExpressionMappingBuilder.cs:71]: This is not covered by tests and I can't think of a use-case, is it really required?

File: src/Riok.Mapperly/Descriptors/MappingBuilders/InlineExpressionMappingBuilder.cs [hunk 4]
```diff
@@ -25,21 +25,51 @@ public static class InlineExpressionMappingBuilder
         }
 
         var methodSyntax = methodSyntaxRef.GetSyntax();
-        if (methodSyntax is not MethodDeclarationSyntax { ExpressionBody: { } body, ParameterList.Parameters: [var sourceParameter] })
+
+        var bodyExpression = methodSyntax switch
```
Comment (latonz) [inline src/Riok.Mapperly/Descriptors/MappingBuilders/InlineExpressionMappingBuilder.cs:29]: The documentation `queryable-projections.mdx` needs to be updated.

Comment (latonz) [inline src/Riok.Mapperly/Descriptors/MappingBuilders/InlineExpressionMappingBuilder.cs:29]: This adds quite a bit of complexity to this method, I think the readability could be improved a lot by extracting this into another method.

**Review by latonz — CHANGES_REQUESTED** [review]
Thank you for this contribution!
As @trejjam pointed out this works already by using method expressions. I can understand that there are use-cases where one doesn't want to use expressions and I think it's nice that the Mapperly inlining just works out of the box 😊
I added my feedback.

**Comment by latonz** [comment]
Relates https://github.com/riok/mapperly/issues/1406

File: src/Riok.Mapperly/Descriptors/MappingBuilders/InlineExpressionMappingBuilder.cs (hunk 1 above)
Comment (hartmair) [reply src/Riok.Mapperly/Descriptors/MappingBuilders/InlineExpressionMappingBuilder.cs:32]: Oops, this is not need at all (copy/paste left-over).

File: src/Riok.Mapperly/Descriptors/MappingBuilders/InlineExpressionMappingBuilder.cs (hunk 2 above)
Comment (hartmair) [reply src/Riok.Mapperly/Descriptors/MappingBuilders/InlineExpressionMappingBuilder.cs:43]: thx

File: src/Riok.Mapperly/Descriptors/MappingBuilders/InlineExpressionMappingBuilder.cs (hunk 3 above)
Comment (hartmair) [reply src/Riok.Mapperly/Descriptors/MappingBuilders/InlineExpressionMappingBuilder.cs:71]: `UserImplementedInlinedExpressionMapping.ReplaceSource` needs `sourceParameter` later. I moved this up a bit in order to save redundant `MethodDeclarationSyntax` checks.

**Review by latonz — APPROVED** [review]

**Pull request merged**